invoke/runners.py

```diff
@@ -322,8 +322,22 @@ def get():
                     # Can't use six.b because that just assumes latin-1 :(
                     data = data.encode(self.encoding)
                 yield data
-        # Decode stream using our generator & requested encoding
-        for data in codecs.iterdecode(get(), self.encoding, errors='replace'):
+        if not hasattr(output, 'encoding') or self.encoding == output.encoding:
+            # No need in re-encoding if output is a BytesIO or CarbonCopy, or
+            # source and target encodings are the same
+            data_stream = get()
+        else:
+            # Re-encode from user requested encoding to the output encoding
+            data_stream = codecs.iterencode(
+                codecs.iterdecode(
+                    get(),
+                    self.encoding,
+                    errors='replace'
+                ),
+                output.encoding,
+                errors='replace'
+            )
+        for data in data_stream:
```

Comment:
This seems wrong. The original code is pulling ```data``` from the results of ```iterdecode``` - i.e., ```data``` is a string. The replacement code is pulling ```data``` from ```iterencode``` or ```get```, so it's bytes.

This would be the sort of error that Python 2 lets you make because of its incomplete separation of bytes and strings. But it should fail on Python 3.